Fix CSSSelectorParser.cpp dependency on CSSSelectorInlines.h

mcatanzaro opened the pull request:
#### d9cb0beddac4f8a15481128a8fd2e262b46c7a9f
<pre>
Fix CSSSelectorParser.cpp dependency on CSSSelectorInlines.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=266998">https://bugs.webkit.org/show_bug.cgi?id=266998</a>

Reviewed by Darin Adler.

Also, move an unrelated dependency to a more appropriate location in the
CMake file.

* Source/WebCore/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/272634@main">https://commits.webkit.org/272634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85c019e47ee83cd7ffebb1cab008eb8d35d49fc3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32231 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10966 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34046 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34790 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13317 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8175 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28776 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32649 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9247 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28815 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8059 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8206 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28734 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36134 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29307 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29180 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34336 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8329 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6265 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32200 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9971 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7558 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8962 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8869 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2556,7 +2556,7 @@ list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/SelectorPseudoClassAn
 ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelector.cpp CSSSelectorEnums.h)
 ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelector.cpp CSSSelectorInlines.h)
-ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelectorParser.cpp CSSSelectorInlines.h)
+ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/parser/CSSSelectorParser.cpp CSSSelectorInlines.h)
```
If a `cpp` file includes another file isn't that dependency detected automatically when the file is compiled? Maybe this is only needed to trigger the code generation the first time?

nt1m replied:
(Note that I have no idea if this was necessary to add, I just copied the pattern that `CSSPropertyParsing.h` was using here, it's possible that it's not needed at all)

mcatanzaro replied:
> If a cpp file includes another file isn't that dependency detected automatically when the file is compiled?

When deciding whether the file needs to be recompiled, yes. Not when deciding whether the file needs to be regenerated.

But you're right: these are not generated files, so it is strange for them to manually specify dependencies on particular headers. Maybe it's a mistake for these dependencies to exist at all. I will take a quick look tomorrow and will probably remove these if I can't think of some reason for this.

mcatanzaro replied:
> But you're right: these are not generated files, so it is strange for them to manually specify dependencies on particular headers. Maybe it's a mistake for these dependencies to exist at all. I will take a quick look tomorrow and will probably remove these if I can't think of some reason for this.

Sorry, what I wrote here yesterday makes no sense. Build systems _do_ automatically consider whether header dependencies have been touched when deciding whether a source file needs to be recompiled for an incremental build. But they do not ensure the generated files exist before attempting to compile the source files that require them. It's normal for non-generated source files to depend on generated headers, and specifying the dependency manually is required to enforce the build order.

That said, there are some other problems that I missed yesterday:

 * Normally we only need to specify dependencies on header files, because it would be very strange for a source file to fail to build if another source file has not yet been built, because source files _normally_ only include header files. So the dependency of HTMLTreeBuilder.cpp on MathMLNames.cpp is really weird. Similarly, it's weird that StyleResolver.cpp and UserAgentStyle.cpp both depend on UserAgentStyleSheetsData.cpp. I've investigated and found these particular dependencies are bogus and should just be removed. (But, for example, it's not weird that StyleResolver.cpp and UserAgentStyle.cpp depend on UserAgentStyleSheets.h. That's a real dependency.)
 * The source file that depends on other files must be excluded from unified build, or the dependency does not work because the source file is never directly compiled by the build system. (It's not possible to make unified source files depend on other files.) But every source file passed to ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES() is part of the unified build, so every such dependency is already broken and does nothing. We need to exclude affected files CSSParser.cpp, CSSSelector.cpp, CSSSelectorParser.cpp, Document.cpp, StyleResolver.cpp, and UserAgentStyle.cpp from the unified build (which is everything that's left after removing the bogus deps). I'm not sure if I'll be able to do this myself or not; modifying XCode files without XCode is uncomfortable, so I might need to ask for help here. (We'll see.)

mcatanzaro replied:
I'm going to land these changes now anyway, and will follow up on this via https://bugs.webkit.org/show_bug.cgi?id=266984.

mcatanzaro replied:
Actually, according to https://cmake.org/cmake/help/latest/prop_sf/GENERATED.html:

> When a generated file created as the OUTPUT of an add_custom_command() command is explicitly listed as a source file for any target in the same directory scope (which usually means the same CMakeLists.txt file), CMake will automatically create a dependency to make sure the file is generated before building that target.

That's nice, but unfortunately we still cannot rely on that because of unified sources.

mcatanzaro committed:
Fix CSSSelectorParser.cpp dependency on CSSSelectorInlines.h
https://bugs.webkit.org/show_bug.cgi?id=266998

Reviewed by Darin Adler.

Also, move an unrelated dependency to a more appropriate location in the
CMake file.

* Source/WebCore/CMakeLists.txt:

Canonical link: https://commits.webkit.org/272634@main
